File: Java/src/org/hpms/mw/distcrud/Cache.java
package org.hpms.mw.distcrud;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Collection;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import org.hpms.dbg.Performance;

final class Cache implements ICache {

   private static byte _NextCacheId = 1;

   private final Set<ClassID>         _classes        = new HashSet<>();
   private final Set<Shareable>       _updated        = new HashSet<>();
   private final Set<Shareable>       _deleted        = new HashSet<>();
   private final Set<ByteBuffer>      _toUpdate       = new HashSet<>();
   private final Set<GUID>            _toDelete       = new TreeSet<>();
   private final Map<GUID, Shareable> _local          = new TreeMap<>();
   private /* */ int                  _nextInstance   = 1;
   private /* */ boolean              _ownershipCheck = false;
   private final ParticipantImpl      _participant;
   private final short                _publisherId;
   /*   */ final byte                 _cacheId;

   Cache( ParticipantImpl participant ) {
      _participant = participant;
      _publisherId = participant._publisherId;
      _cacheId     = _NextCacheId++;
   }

   boolean matches( short publisherId, byte cacheId ) {
      return( publisherId == _publisherId ) && ( cacheId == _cacheId );
   }

   @Override
   public boolean owns( GUID id ) {
      return matches( id._publisher, id._cache );
   }

   @Override
   public void setOwnership( boolean enabled ) {
      _ownershipCheck = enabled;
   }

   @Override
   public Status create( Shareable item ) {
      if( item._id.isShared()) {
         return Status.ALREADY_CREATED;
      }
      item._id._publisher = _publisherId;
      item._id._cache     = _cacheId;
      item._id._instance  = _nextInstance++;
      synchronized( _local ) {
         _local.put( item._id, item );
      }
      synchronized( _updated ) {
         _updated.add( item );
      }
      return Status.NO_ERROR;
   }

   @SuppressWarnings("unchecked")
   @Override
   public <T extends Shareable> T read( GUID id ) {
      return (T)_local.get( id );
   }

   @Override
   public Status update( Shareable item ) {
      if( _ownershipCheck && ! owns( item._id )) {
         return Status.NOT_OWNER;
      }
      if( item._id._instance == 0 ) {
         return Status.NOT_CREATED;
      }
      if( ! _local.containsKey( item._id )) {
         return Status.NOT_IN_THIS_CACHE;
      }
      synchronized( _updated ) {
         _updated.add( item );
      }
      return Status.NO_ERROR;
   }

   @Override
   public Status delete( Shareable item ) {
      synchronized( _local ) {
         if( _ownershipCheck && ! owns( item._id )) {
            return Status.NOT_OWNER;
         }
         if( _local.remove( item._id ) == null ) {
            return Status.NOT_IN_THIS_CACHE;
         }
         synchronized( _deleted ) {
            _deleted.add( item );
         }
      }
      return Status.NO_ERROR;
   }

   @Override
   public Collection<Shareable> values() {
      return _local.values();
   }

   @Override
   public Set<Shareable> select( Predicate<Shareable> query ) {
      return _local.values()
         .parallelStream()
         .filter( query )
         .collect( Collectors.toSet());
   }

   @Override
   public void publish() throws IOException {
      final long atStart = System.nanoTime();
      synchronized( _updated ) {
         _participant.publishUpdated( _updated );
         _updated.clear();
      }
      synchronized( _deleted ) {
         _participant.publishDeleted( _deleted );
         _deleted.clear();
      }
      Performance.record( "publish", System.nanoTime() - atStart );
   }

   @Override
   public void subscribe( ClassID id ) {
      _classes.add( id );
   }

   @Override
   public void refresh() {
      final long atStart = System.nanoTime();
      synchronized( _local ) {
         synchronized( _toUpdate ) {
            for( final ByteBuffer update : _toUpdate ) {
               final GUID      id      = GUID.unserialize( update );
               final ClassID   classId = ClassID.unserialize( update );
               final Shareable t       = _local.get( id );
               if( t == null ) {
                  final Shareable item = _participant.newInstance( classId, update );
                  if( item != null ) {
                     item._id.set( id );
                     _local.put( id, item );
                  }
                  else {
                     System.err.printf( "Unknown %s of %s\n", classId, id );
                  }
               }
               else if( ! _ownershipCheck || ! owns( id )) {
                  t.unserialize( update );
               }
            }
            _toUpdate.clear();
         }
         synchronized( _toDelete ) {
            _local.keySet().removeAll( _toDelete );
            _toDelete.clear();
         }
      }
      Performance.record( "refresh", System.nanoTime() - atStart );
   }

   void updateFromNetwork( ByteBuffer source ) {
      synchronized( _toUpdate ) {
         _toUpdate.add( source );
      }
   }

   void deleteFromNetwork( GUID id ) {
      synchronized( _toDelete ) {
         _toDelete.add( id );
      }
   }
}

File: cpp/src/dcrud/Cache.cpp
#include "Cache.hpp"

#include <dcrud/Shareable.hpp>

#include <stdio.h>

using namespace dcrud;

#ifdef NO_ERROR
#  undef NO_ERROR
#endif

Cache::Cache( ParticipantImpl & participant ) :
   _nextInstance  ( 1U          ),
   _ownershipCheck( false       ),
   _participant   ( participant )
{}

Cache::~ Cache() {
}

bool Cache::owns( const dcrud::GUID & id ) const {
   return _participant.getPublisherId() == id._publisher;
}

Status Cache::create( Shareable & item ) {
   if( item._id.isShared()) {
      return ALREADY_CREATED;
   }
   item._id._publisher = _participant.getPublisherId();
   item._id._instance  = _nextInstance++;
   {
      os::Synchronized sync( _localMutex );
      _local[item._id] = &item;
   }
   {
      os::Synchronized sync( _updatedMutex );
      _updated.insert( &item );
   }
   return NO_ERROR;
}

Shareable * Cache::read( const dcrud::GUID & id ) const {
   localCstIter_t it = _local.find( id );
   if( it == _local.end()) {
      return 0;
   }
   return it->second;
}


Status Cache::update( Shareable & item ) {
   if( _ownershipCheck && ! owns( item._id )) {
      return NOT_OWNER;
   }
   if( item._id._instance == 0 ) {
      return NOT_CREATED;
   }
   {
      os::Synchronized sync( _localMutex );
      localCstIter_t it = _local.find( item._id );
      if( it == _local.end()) {
         return NOT_IN_THIS_CACHE;
      }
   }
   {
      os::Synchronized sync( _updatedMutex );
      _updated.insert( &item );
   }
   return NO_ERROR;
}

Status Cache::deleTe( Shareable & item ) {
   if( _ownershipCheck && ! owns( item._id )) {
      return NOT_OWNER;
   }
   {
      os::Synchronized sync( _localMutex );
      if( _local.erase( item._id ) < 1 ) {
         return NOT_IN_THIS_CACHE;
      }
   }
   {
      os::Synchronized sync( _deletedMutex );
      _deleted.insert( &item );
   }
   return NO_ERROR;
}

void Cache::values( shareables_t & snapshot ) const {
   os::Synchronized sync( _localMutex );
   snapshot.clear();
   for( localCstIter_t it = _local.begin(); it != _local.end(); ++it ) {
      snapshot.insert( it->second );
   }
}

bool Cache::select( shareablePredicate_t query, shareables_t & snapshot ) const {
   os::Synchronized sync( _localMutex );
   snapshot.clear();
   for( localCstIter_t it = _local.begin(); it != _local.end(); ++it ) {
      if( query( *it->second )) {
         snapshot.insert( it->second );
      }
   }
   return ! snapshot.empty();
}

void Cache::publish( void ) {
   {
      os::Synchronized sync( _updatedMutex );
      _participant.publishUpdated( _updated );
      _updated.clear();
   }
   {
      os::Synchronized sync( _deletedMutex );
      _participant.publishDeleted( _deleted );
      _deleted.clear();
   }
}

void Cache::subscribe( const ClassID & id ) {
   _classes.insert( id );
}

void Cache::refresh( void ) {
   {
      os::Synchronized sync( _localMutex );
      {
         os::Synchronized sync2( _toUpdateMutex );
         for( byteBuffersIter_t it = _toUpdate.begin(); it != _toUpdate.end(); ++it ) {
            io::ByteBuffer * update   = *it;
            dcrud::GUID             id       = dcrud::GUID   ::unserialize( *update );
            ClassID          classId  = ClassID::unserialize( *update );
            localIter_t      itemIter = _local.find( id );
            if( itemIter == _local.end()) {
               Shareable * item = _participant.newInstance( classId, *update );
               if( item ) {
                  item->_id.set( id );
                  _local[id] = item;
               }
               else {
                  fprintf( stderr, "Unknown %s of %s\n",
                     classId.toString().c_str(), id.toString().c_str());
               }
            }
            else if( ! _ownershipCheck || ! owns( id )) {
               itemIter->second->unserialize( *update );
            }
            delete update;
         }
         _toUpdate.clear();
      }
      {
         os::Synchronized sync3( _toDeleteMutex );
         for( guidsIter_t it = _toDelete.begin(); it != _toDelete.end(); ++it ) {
            _local.erase( *it );
         }
         _toDelete.clear();
      }
   }
}
void Cache::updateFromNetwork( io::ByteBuffer * source ) {
   os::Synchronized sync( _toUpdateMutex );
   _toUpdate.insert( source );
}

void Cache::deleteFromNetwork( const dcrud::GUID & id ) {
   os::Synchronized sync( _toDeleteMutex );
   _toDelete.insert( id );
}
